io/slow5/slow5.go

```diff
@@ -0,0 +1,409 @@
+
+Slow5 write takes in a header, a channel of reads, and an io.Writer output.
+The intended use case of slow5 write is reading from a location, such as a
+database, and then directly writing the output to somewhere (stdout, a file,
+etc).
+
+A channel is used here so that reading and writing of slow5 can be done
+concurrently. In almost all cases, you do not want to have slow5 files entirely
+in memory, because they're freakin' huge.
+
+Cheers,
+
+Keoni
+
+******************************************************************************/
+
+// Write writes a list of headers and a channel of reads to an output.
+func Write(headers []Header, reads <-chan Read, output io.Writer) error {
+	// First, write the slow5 version number
+	slow5Version := headers[0].Slow5Version
+	_, err := output.Write([]byte(fmt.Sprintf("#slow5_version\t%s\n", slow5Version)))
+	if err != nil {
+		return err
+	}
+	// Then, write the number of read groups (ie, the number of headers)
+	_, err = output.Write([]byte(fmt.Sprintf("#num_read_groups\t%d\n", len(headers))))
+	if err != nil {
+		return err
+	}
+	// Next, we need a map of what attribute values are available
+	possibleAttributeKeys := make(map[string]bool)
+	for _, header := range headers {
+		for key := range header.Attributes {
+			possibleAttributeKeys[key] = true
+		}
+	}
+	// Now that we know what attribute values are possible, lets build a map
+	// with those values with "." placeholders (as defined in slow5 spec)
+	headerAttributes := make(map[string][]string)
+	for key := range possibleAttributeKeys {
+		newBlankAttributes := make([]string, len(headers))
+		for blankAttributeIndex := range newBlankAttributes {
+			newBlankAttributes[blankAttributeIndex] = "."
+		}
+		headerAttributes[key] = newBlankAttributes
+	}
+	// Build a list with all header values
+	var headerAttributeStrings []string
+	for headerIndex, header := range headers {
+		for key, value := range header.Attributes {
+			headerAttributes[key][headerIndex] = value
+		}
+	}
+	for key, value := range headerAttributes {
+		var attributeRow strings.Builder
+		attributeRow.Write([]byte(key))
+		for _, attributeValue := range value {
+			attributeRow.Write([]byte("\t"))
+			attributeRow.Write([]byte(attributeValue))
+		}
+		headerAttributeStrings = append(headerAttributeStrings, attributeRow.String())
+	}
+	// Sort the header strings
+	sort.Strings(headerAttributeStrings)
+
+	// Write the header attribute strings to the output
+	for _, headerAttributeString := range headerAttributeStrings {
+		_, err = output.Write([]byte(fmt.Sprintf("%s\n", headerAttributeString)))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Write the read headers
+	// These are according to the slow5 specifications
+	_, err = output.Write([]byte("#char*	uint32_t	double	double	double	double	uint64_t	int16_t*	uint64_t	int32_t	uint8_t	double	enum{unknown,partial,mux_change,unblock_mux_change,data_service_unblock_mux_change,signal_positive,signal_negative}	char*\n"))
```

Comment:
Might be a good idea to also apply `fmt.Fprintln` here. The reason being is that I'm not sure if a heap allocation is done for the string- I've asked on the performance channel in the gophers slack to check...